core/domain/stats_services.py

```diff
@@ -193,117 +139,99 @@ def get_visualizations_info(exploration_id, exploration_version, state_name):
     """
     return get_top_state_rule_answers_multi(
-        [(exploration_id, state_name)], rule_str_list)[0]
+        [(exploration_id, state_name)], classify_category_list)[0]
 
 
-def get_top_state_rule_answers_multi(exploration_state_list, rule_str_list):
-    """Returns a list of top answers (by submission frequency) submitted to the
-    given explorations and states which were mapped to any of the rules listed
-    in 'rule_str_list' for each exploration ID and state name tuple in
-    exploration_state_list.
+def get_top_state_rule_answers_multi(
+        exploration_state_list, classify_category_list):
+    """Returns a list of top answers (sorted by submission frequency) submitted
+    to the given explorations and states which were mapped to any of the rule
+    classification categories listed in 'classify_category_list'.
 
-    For each exploration ID and state, all answers submitted that match any of
-    the rule spec strings in rule_str_list are returned.
+    NOTE TO DEVELOPERS: Classification categories are stored upon answer
+    submission, so the answers returned by this function may be stale and not
+    evaluate in the same way as they did upon submission.
     """
-    answer_log_list = (
-        stats_domain.StateRuleAnswerLog.get_multi_by_multi_explorations(
-            exploration_state_list, rule_str_list))
-    return [[
-        {
-            'value': top_answer[0],
-            'count': top_answer[1]
-        }
-        for top_answer in answer_log.get_all_top_answers()
-    ] for answer_log in answer_log_list]
+    # TODO(bhenning): This should have a custom, continuous job (possibly as
+    # part of the summarizers framework) which goes through all answers, finds
+    # those which are not covered by hard rules or are not part of the training
+    # data of soft rules, rank them by their frequency, then output them. This
+    # output will have reasonably up-to-date answers which need to be resolved
+    # by creators.
+    answer_lists = []
+    for exploration_id, state_name in exploration_state_list:
+        # pylint: disable=line-too-long
+        job_result = (
+            stats_jobs_continuous.InteractionAnswerSummariesAggregator.get_calc_output(
+                exploration_id, state_name, 'TopAnswersByCategorization'))
+        if job_result:
+            calc_output = job_result.calculation_output
+            answer_list = list(itertools.chain.from_iterable(
+                calc_output[category]
+                for category in classify_category_list
+                if category in calc_output))
+
+            # If the answer_list includes similar answers matching multiple
+            # categories, those answers should be de-duplicated.
+            # TODO(bhenning): Make this better than O(N^2); probably better just
+            # to implement the job described above.
+            combined_answer_list = [
+                {'answer': answer['answer'], 'frequency': 0}
```

Comment:
#3289 this is important but I worry about the risk of introducing regressions since so many places need to change. We should clean this up, but I don't think before the migration is the right time to do so.